modules/afsocket/afsocket-source.c

```diff
@@ -658,33 +684,88 @@ _dynamic_window_timer_stop(AFSocketSourceDriver *self)
   if (iv_timer_registered(&self->dynamic_window_timer))
     iv_timer_unregister(&self->dynamic_window_timer);
 }
+
+
+#ifdef SYSLOG_NG_ENABLE_PACKET_STATS
+
 static void
-_listen_fd_stop(AFSocketSourceDriver *self)
+_packet_stats_timer_start(AFSocketSourceDriver *self)
 {
-  if (iv_fd_registered (&self->listen_fd))
-    iv_fd_unregister(&self->listen_fd);
+  if (!self->stat_socket_dropped_packets)
+    return;
+
+  iv_validate_now();
+  self->packet_stats_timer.expires = iv_now;
+  timespec_add_msec(&self->packet_stats_timer.expires, PACKET_STATS_TIMER_MSECS);
+  iv_timer_register(&self->packet_stats_timer);
+}
+
+
+static void
+_on_packet_stats_timer_elapsed(gpointer cookie)
+{
+  AFSocketSourceDriver *self = (AFSocketSourceDriver *)cookie;
+  guint32 meminfo[SK_MEMINFO_VARS];
+  socklen_t meminfo_len = sizeof(meminfo);
+  AFSocketSourceConnection *first_connection = self->connections->data;
+  gint sock = first_connection->sock;
+
+  /* once this getsockopt() fails, we won't try again */
+  if (getsockopt(sock, SOL_SOCKET, SO_MEMINFO, &meminfo, &meminfo_len) < 0)
+    return;
+
+  stats_counter_set(self->stat_socket_dropped_packets, meminfo[SK_MEMINFO_DROPS]);
+  stats_counter_set(self->stat_socket_receive_buffer_max, meminfo[SK_MEMINFO_RCVBUF]);
+  stats_counter_set(self->stat_socket_receive_buffer_used, meminfo[SK_MEMINFO_RMEM_ALLOC]);
+  _packet_stats_timer_start(self);
+}
+
+static void
+_packet_stats_timer_init(AFSocketSourceDriver *self)
+{
+  IV_TIMER_INIT(&self->packet_stats_timer);
+  self->packet_stats_timer.cookie = self;
+  self->packet_stats_timer.handler = _on_packet_stats_timer_elapsed;
+}
+
+
+static void
+_packet_stats_timer_stop(AFSocketSourceDriver *self)
+{
+  if (iv_timer_registered(&self->packet_stats_timer))
+    iv_timer_unregister(&self->packet_stats_timer);
 }
 
+#else
+
+#define _packet_stats_timer_init(s)
+#define _packet_stats_timer_start(s)
+#define _packet_stats_timer_stop(s)
+
+#endif
+
 static void
 _init_watches(AFSocketSourceDriver *self)
 {
   _dynamic_window_timer_init(self);
+  _packet_stats_timer_init(self);
+  _listen_fd_init(self);
 }
 
 static void
 afsocket_sd_start_watches(AFSocketSourceDriver *self)
 {
   _listen_fd_start(self);
-
-  if (self->dynamic_window_pool != NULL)
-    _dynamic_window_timer_start(self);
+  _dynamic_window_timer_start(self);
+  _packet_stats_timer_start(self);
```

Comment:
We'd probably need an immediate `_on_packet_stats_timer_elapsed()` call here if we make the timer interval configurable (so we see max stats at startup/reload quickly).